### PR TITLE
Implement batched and offline metrics

### PR DESCRIPTION
**Pull request by ianwremmel**

### CODE REVIEW

File: packages/plugin-metrics/test/unit/spec/batcher.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+import Metrics, {config} from '../..';
+import MockSpark from '@ciscospark/test-helper-mock-spark';
+import sinon from '@ciscospark/test-helper-sinon';
+import {SparkHttpError} from '@ciscospark/spark-core';
+
+function promiseTick(count) {
+  let promise = Promise.resolve();
+  while (count > 1) {
+    promise = promise.then(() => {
+      return promiseTick(1);
```
Comment (good-joe) [inline packages/plugin-metrics/test/unit/spec/batcher.js]: Why the recursion? Doesn't promiseTick just return Promise.resolve() if count === 1?Sorry if I'm missing something

File: packages/plugin-metrics/test/unit/spec/batcher.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+              body: undefined,
+              options
+            });
+          });
+
+          const promise = spark.metrics.batcher.request({
+            key: `testMetric`
+          });
+
+          return promiseTick(50)
```
Comment (good-joe) [inline packages/plugin-metrics/test/unit/spec/batcher.js]: nice!

**Comment by good-joe** [comment]
One question on style rather than substance

LGTM

File: packages/plugin-metrics/test/unit/spec/batcher.js (hunk 1 above)
Comment (ianwremmel) [inline packages/plugin-metrics/test/unit/spec/batcher.js]: if i'm understanding your question correctly, that would produce a very short synchronous wait. the goal of promiseTick is to let `count` ticks pass by chaining `count` `promise.then`s.

basically, it's a hack to deal with the fact that lolex replaces the clock used by setTimeout and setInterval, but not the clock used by promises.

**Pull request merged**